docs/source/extension/plugin_system.rst

```diff
@@ -0,0 +1,179 @@
+    activate: (app: JupyterFrontEnd) => {
+      console.log('JupyterLab extension (step_counter/provider plugin) is activated!');
+      const counter = new Counter();
+
+      // Since this plugin "provides" the "StepCounter" service, make sure to
+      // return the object you want to use as the "service object" here (when
+      // other plugins request the StepCounter service, it is this object
+      // that will be supplied)
+      return counter;
+    }
+  };
+
+Here, you can see that this plugin lists a "StepCounter" token object as its
+"provides" property, which tells JupyterLab that it is a "provider" of a
+service object.
+
+It also returns a "Counter" instance: this is the service object it "provides"
+for the StepCounter service.
+
+When your plugin becomes a provider, you need to define a lumino "Token" object
+that JupyterLab will use to identify the service. Here's how the StepCounter
+Token was defined:
+
+.. code::
+
+  // The token is used to identify a particular "service" in
+  // JupyterLab's extension system (here the StepCounter token
+  // identifies the example "Step Counter Service", which is used
+  // to store and increment step count data in JupyterLab). Any
+  // plugin can use this token in their "requires" or "activates"
+  // list to request the service object associated with this token!
+  const StepCounter = new Token<StepCounterItem>(
+    'step_counter:StepCounter',
+    'A service for counting steps.'
+  );
+
+Note that StepCounter is a Lumino Token object. The StepCounter defined
+here also passes the "StepCounterItem" interface in the Token definition.
+
+When you provide an interface to your Token definition in this way, you're
+telling JupyterLab to type check the service object it gets from any provider
+plugin associated with this service, to make sure it conforms to that
+interface. This helps ensure that any provider plugin (even a substitute
+provider that someone else makes later) provides a compatible service object
+(in this case, a StepCounterItem object), and it helps enable the plugin
+swappability and subsitution in JupyterLab.
+
+Here's the interface the token uses:
+
+.. code::
+
+  // The StepCounterItem interface is used as part of JupyterLab's
+  // provider-consumer pattern. This interface is supplied to the
+  // token instance (the StepCounter token), and JupyterLab will
+  // use it to type-check any service-object associated with the
+  // token that a provider plugin supplies to check that it conforms
+  // to the interface.
+  interface StepCounterItem {
+    // registerStatusItem(id: string, statusItem: IStatusBar.IItem): IDisposable;
+    getStepCount(): number;
+    incrementStepCount(count: number): void;
+    countChanged: Signal<any, number>;
+  }
+
+This means that anyone who makes a provider plugin for the StepCounter service
+must return an object that has a getStepCount method, incrementStepCount method,
+and a countChanges Signal (a Lumino Signal object).
```

Comment:
If we are creating a separate doc, it would be good to cover
- `deactivate()`
- mention naming conventions for identifiers
- check if we need to expose anything else from lumino's `IPlugin`
- deferred plugins
- maybe `async activate` and what it implies

Reply:
I can add these, probably in a separate PR 👍

Reply:
The plugin system doc has been split out [here](https://github.com/jupyterlab/jupyterlab/pull/15417).